caldav: get calendar data using PROPFIND

voidcontext opened the pull request:
This change adds FindCalendar calendar to the caldav client. It can be
used to get calendar information without loading all the events once the
path of the calendar is known.

emersion commented:
This is too specific for a particular use-case.

Could we return a full `Calendar` instead?

voidcontext commented:
> This is too specific for a particular use-case.
> 
> Could we return a full `Calendar` instead?

That's fair, let me have a look.

voidcontext committed:
caldav: get calendar data using PROPFIND

This change adds FindCalendar function to the caldav client. It can be
used to get calendar information without loading all the events, once the
path of the calendar is known.

voidcontext commented:
@emersion I've updated the PR, this should be more general, I wasn't sure about the naming though. What do you think?

emersion commented on caldav/client.go:
```diff
@@ -42,6 +42,31 @@ func NewClient(c webdav.HTTPClient, endpoint string) (*Client, error) {
+// Get calendar information without loading all the events once the path of the calendar is known.
+// As path use
+// - ${calId}/user (Google)
+// - calendar path (standard implementations)
```
Nit: can we leave out any provider-specific details here?

voidcontext replied:
Removed.

emersion commented on caldav/client.go:
```diff
@@ -42,6 +42,31 @@ func NewClient(c webdav.HTTPClient, endpoint string) (*Client, error) {
+// As path use
+// - ${calId}/user (Google)
+// - calendar path (standard implementations)
+func (c *Client) FindCalendar(ctx context.Context, path string) (Calendar, error) {
```
Could we make this function return a `*Calendar` instead of a `Calendar`?

- We can return `nil` on error, instead of an empty struct
- The struct has quite a few fields, returning a pointer avoids copying
- A pointer is more idiomatic

voidcontext replied:
I am fairly new to go, and I still find it difficult sometimes to decide between returning a pointer or the struct. It seems I need to be more aggressive, as I wouldn't have considered the Calendar struct to have enough fields that'd warrant  a pointer. Thanks for sharing your insights!

emersion commented on caldav/client.go:
```diff
@@ -42,6 +42,31 @@ func NewClient(c webdav.HTTPClient, endpoint string) (*Client, error) {
+// - ${calId}/user (Google)
+// - calendar path (standard implementations)
+func (c *Client) FindCalendar(ctx context.Context, path string) (Calendar, error) {
+	propfind := internal.NewPropNamePropFind(
```
Nit: we could extract this into a global `var` and use it in the two functions performing a `PROPFIND` on a calendar. This would ensure we always request the information nee by `decodeCalendar` from the server.

voidcontext replied:
Actually I was thinking about doing this!

emersion commented on caldav/client.go:
```diff
@@ -87,44 +112,53 @@ func (c *Client) FindCalendars(ctx context.Context, calendarHomeSet string) ([]C
-		if err := resp.DecodeProp(&supportedCompSet); err != nil && !internal.IsNotFound(err) {
-			return nil, err
-		}
+func decodeCalendar(path string, resp *internal.Response) (Calendar, error) {
```
Ditto: returning `*Calendar` would be nicer here

voidcontext replied:
Refactored.

emersion reviewed (commented):
Can you edit the commit and PR title to not use conventional commits? We use the package name as the prefix, just like the Go repository does.

> I wasn't sure about the naming though. What do you think?

Right! `FindCalendar` is consistent with the existing functions, but is perhaps a bit too easy to mix up with `FindCalendars` (only one letter differs). This probably calls for a wider renaming - we can leave this up for a future issue/PR.

emersion reviewed (approved):
Thank you!
